pkg/tlsutil/tls.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2018 The Operator-SDK Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlsutil
+
+import (
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// CertType defines the type of the cert.
+type CertType int
+
+const (
+	// ClientCert defines a client cert.
+	ClientCert CertType = iota
+	// ServingCert defines a serving cert.
+	ServingCert
+	// ClientAndServingCert defines both client and serving cert.
+	ClientAndServingCert
+)
+
+// CertConfig configures how to generate the Cert.
+type CertConfig struct {
+	// CertName is the name of the cert.
+	CertName string
+	// Optional CertType. Serving, client or both; defaults to both.
+	CertType CertType
+	// Optional CommonName is the common name of the cert; defaults to "".
+	CommonName string
+	// Optional Organization is Organization of the cert; defaults to "".
+	Organization []string
+	// Optional CA Key, if user wants to provide custom CA key via a file path.
+	CAKey string
+	// Optional CA Certificate, if user wants to provide custom CA cert via file path.
+	CACert string
+}
+
+// CertGenerator is an operator specific TLS tool that generates TLS assets for the deploying application.
+type CertGenerator interface {
+	// GenerateCert generates a secret containing TLS encryption key and cert, a Secret
+	// containing the CA key, and a ConfigMap containing the CA Certificate given the Custom
+	// Resource(CR) "cr", the Kubernetes Service "Service", and the CertConfig "config".
+	//
+	// The passed in "service" represents endpoint(s) of a communicating party. We assume that the
+	// deployed applications are typically communicated via the Kubernete Service. Therefore, the
+	// cert is generated to certify the endpoint(s) defined by the "service".
+	//
+	// If CA is not passed in via CertConfig, GenerateCert creates a unique self signed CA
+	// corresponding to the "cr" for signing the TLS cert and also puts the self signed CA and TLS
+	// key and Certs into the Kubernetes cluster before returning.
+	// If CA is passed in, GenerateCert use it for signing the TLS cert but
+	// only puts TLS key and cert into the cluster not the passed in CA; the user has to
+	// figure out what to do with the returned CA object.
```

Comment:
when you say "the user has to figure out what to do with the CA object" do you mean they can re-store it where they like?

Reply:
kinda of. since when you pass-in the CA, we return a CA configmap and secret back to you. However, we are not going to puts them into the kubernetes. It is the user's job to figure out whether to put them into kubernete or not.